### PR TITLE
Add plugin helpers for parser and formatter

### PR DESCRIPTION
**Pull request by tagomoris**

### CODE REVIEW

**Comment by tagomoris** [comment]
@repeatedly please review this change.

File: lib/fluent/plugin_helper/formatter.rb [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+            block.call(formatter) if block_given?
+          rescue => e
+            log.error "unexpected error while #{method_name}", usage: usage, formatter: formatter, error: e
+          end
+        end
+      end
+
+      def stop
+        super
+        # timer stops automatically in super
```
Comment (repeatedly) [inline lib/fluent/plugin_helper/formatter.rb]: What does this comment mean?

File: lib/fluent/plugin_helper/formatter.rb [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+
+require 'monitor'
+require 'forwardable'
```
Comment (repeatedly) [inline lib/fluent/plugin_helper/formatter.rb]: Are these `require needed?

File: lib/fluent/plugin_helper/parser.rb [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+
+require 'monitor'
+require 'forwardable'
```
Comment (repeatedly) [inline lib/fluent/plugin_helper/parser.rb]: ditto

File: lib/fluent/plugin_helper/formatter.rb [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+module Fluent
+  module PluginHelper
+    module Formatter
+      def formatter_create(usage: '', type: nil, conf: nil)
+        formatter = @_formatters[usage]
+        return formatter if formatter
+
+        if !type
+          raise ArgumentError, "BUG: both type and conf are not specified" unless conf
+          raise Fluent::ConfigError, "@type is not specified for <format>" unless conf['@type']
```
Comment (repeatedly) [inline lib/fluent/plugin_helper/formatter.rb]: For me, `@type is required in <format>` is clear than this message but not critical.

File: lib/fluent/plugin_helper/formatter.rb [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+          end
+          formatter = Plugin.new_formatter(section[:@type], parent: self)
+          formatter.configure(section.corresponding_config_element)
+          @_formatters[section.usage] = formatter
+        end
+      end
+
+      def start
+        super
+        @_formatters_started = true
```
Comment (repeatedly) [inline lib/fluent/plugin_helper/formatter.rb]: This line should be moved to end of this fucntion?
Formatters don't start yet in this position.

File: lib/fluent/plugin_helper/formatter.rb [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+      end
+
+      def close
+        formatter_operate(:close)
+        super
+      end
+
+      def terminate
+        formatter_operate(:terminate)
+        @_formatters = {}
```
Comment (repeatedly) [inline lib/fluent/plugin_helper/formatter.rb]: Is `@_formatters_started = false` needed for complete clean up?

File: lib/fluent/plugin_helper/formatter.rb (hunk 2 above)
Comment (tagomoris) [inline lib/fluent/plugin_helper/formatter.rb]: Ah, I missed to remove these.

File: lib/fluent/plugin_helper/formatter.rb (hunk 4 above)
Comment (tagomoris) [inline lib/fluent/plugin_helper/formatter.rb]: It's understandable for me too.

File: lib/fluent/plugin_helper/formatter.rb (hunk 5 above)
Comment (tagomoris) [inline lib/fluent/plugin_helper/formatter.rb]: No. It's needed to set the variable to true here.
If any formatter calls `formatter_create`, that instance will not be started if we set it true at the end of this method.

File: lib/fluent/plugin_helper/formatter.rb (hunk 1 above)
Comment (tagomoris) [inline lib/fluent/plugin_helper/formatter.rb]: Oops, it should be removed.

File: lib/fluent/plugin_helper/formatter.rb (hunk 6 above)
Comment (tagomoris) [inline lib/fluent/plugin_helper/formatter.rb]: Right.

**Comment by tagomoris** [comment]
I pushed a commit to fix for review comments.

**Pull request merged**